File: rust-lab/src/web_client/utils/web/route_pattern.rs
use std::fmt::{Debug, Formatter};
use crate::web_client::enums::{RequestHandler, PathType};
use crate::web_client::enums::PathType::{DYNAMIC, STATIC};
use crate::web_client::utils::pair::Pair;
pub struct RoutePattern {
    pub(crate) pattern: String,
    pub(crate) path_type: PathType,
    pub(crate) segments: Vec<Pair<String, String>>,
    pub(crate) handler: Option<RequestHandler>,
}

impl RoutePattern {
    pub fn new(pattern: &str, path_type: PathType, segments: Vec<Pair<String, String>>) -> RoutePattern {
        RoutePattern{
            pattern: pattern.to_string(),
            path_type,
            segments,
            handler: None,
        }
    }
    pub fn parse(path: &str) -> RoutePattern {
        let splits:Vec<&str> = path.split("/").collect();
        let mut segments: Vec<Pair<String, String>> = Vec::new();
        let mut variable_exists = false;
        for split in splits {
            if split.starts_with("{") && split.ends_with("}") {
                variable_exists = true;
                segments.push(Pair::new("*".to_string(),split[1..split.len()-1].to_string()));
            }else{
                segments.push(Pair::new(split.to_string(),split.to_string()));
            }
        }
        Self::new(
            path,
             if variable_exists { DYNAMIC } else { STATIC },
            segments
        )
    }
}

impl Debug for RoutePattern {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RoutePattern")
            .field("pattern", &self.pattern)
            .field("path_type", &self.path_type)
            .field("segments", &self.segments)
            .field("handler", &"<handler>")
            .finish()
    }
}

File: labs/lab/src/main/java/org/example/web/utils/web/RoutePattern.java
package org.example.web.utils.web;

import org.example.web.RequestHandler;
import org.example.web.utils.Pair;

import java.util.ArrayList;
import java.util.List;

public class RoutePattern {
    private final String pattern;
    private final PathType pathType;
    private final List<Pair<String, String>> segments;
    private RequestHandler requestHandler;

    private RoutePattern(String pattern, PathType pathType, List<Pair<String, String>> segments) {
        this.pattern = pattern;
        this.pathType = pathType;
        this.segments = segments;
        this.requestHandler = null;
    }

    public static RoutePattern parse(String path){
        String[] split = path.split("/");
        ArrayList<Pair<String, String>> segments = new ArrayList<>();
//        ArrayList<String> variableNames = new ArrayList<>();
        boolean variableExists = false;
        for (String s : split) {
            if(s.startsWith("{") && s.endsWith("}")){
                variableExists = true;
                segments.add(new Pair<>("*", s.substring(1,s.length()-1)));
            }else{
                segments.add(new Pair<>(s, s));
            }
        }
        PathType pathType = variableExists ? PathType.DYNAMIC : PathType.STATIC;
        return new RoutePattern(path, pathType, segments);
    }

    public String getPattern() {
        return pattern;
    }

    public PathType getPathType() {
        return pathType;
    }

    public List<Pair<String, String>> getSegments() {
        return segments;
    }


    public void setRequestHandler(RequestHandler requestHandler) {
        this.requestHandler = requestHandler;
    }

    public RequestHandler getRequestHandler() {
        return requestHandler;
    }

    @Override
    public String toString() {
        return "RoutePattern{" +
                "pattern='" + pattern + '\'' +
                ", pathType=" + pathType +
                ", segments=" + segments +
                ", requestHandler=" + requestHandler +
                '}';
    }
}
